src/coprocessor/codec/raw_datum.rs

```diff
@@ -0,0 +1,298 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+//! Decoders to decode into a concrete datum.
+
+use cop_datatype::{FieldTypeAccessor, FieldTypeTp};
+use tikv_util::codec::{bytes, number};
+use tipb::expression::FieldType;
+
+use super::data_type::*;
+use crate::coprocessor::codec::datum;
+use crate::coprocessor::codec::mysql::Tz;
+use crate::coprocessor::codec::{Error, Result};
+
+#[inline]
+fn decode_int(v: &mut &[u8]) -> Result<i64> {
+    number::decode_i64(v)
+        .map_err(|_| Error::InvalidDataType("Failed to decode data as i64".to_owned()))
+}
+
+#[inline]
+fn decode_uint(v: &mut &[u8]) -> Result<u64> {
+    number::decode_u64(v)
+        .map_err(|_| Error::InvalidDataType("Failed to decode data as u64".to_owned()))
+}
+
+#[inline]
+fn decode_var_int(v: &mut &[u8]) -> Result<i64> {
+    number::decode_var_i64(v)
+        .map_err(|_| Error::InvalidDataType("Failed to decode data as var_i64".to_owned()))
+}
+
+#[inline]
+fn decode_var_uint(v: &mut &[u8]) -> Result<u64> {
+    number::decode_var_u64(v)
+        .map_err(|_| Error::InvalidDataType("Failed to decode data as var_u64".to_owned()))
+}
+
+#[inline]
+fn decode_float(v: &mut &[u8]) -> Result<f64> {
+    number::decode_f64(v)
+        .map_err(|_| Error::InvalidDataType("Failed to decode data as f64".to_owned()))
+}
+
+#[inline]
+fn decode_decimal(v: &mut &[u8]) -> Result<Decimal> {
+    Decimal::decode(v)
+        .map_err(|_| Error::InvalidDataType("Failed to decode data as decimal".to_owned()))
+}
+
+#[inline]
+fn decode_bytes(v: &mut &[u8]) -> Result<Vec<u8>> {
+    bytes::decode_bytes(v, false)
+        .map_err(|_| Error::InvalidDataType("Failed to decode data as bytes".to_owned()))
+}
+
+#[inline]
+fn decode_compact_bytes(v: &mut &[u8]) -> Result<Vec<u8>> {
+    bytes::decode_compact_bytes(v)
+        .map_err(|_| Error::InvalidDataType("Failed to decode data as compact bytes".to_owned()))
+}
+
+#[inline]
+fn decode_json(v: &mut &[u8]) -> Result<Json> {
+    Json::decode(v).map_err(|_| Error::InvalidDataType("Failed to decode data as json".to_owned()))
+}
+
+#[inline]
+fn decode_duration_from_i64(v: i64) -> Result<Duration> {
+    Duration::from_nanos(v, 0)
+        .map_err(|_| Error::InvalidDataType("Failed to decode i64 as duration".to_owned()))
+}
+
+#[inline]
+fn decode_date_time_from_uint(v: u64, time_zone: &Tz, field_type: &FieldType) -> Result<DateTime> {
+    use std::convert::TryInto;
+
+    let fsp = field_type.decimal() as i8;
+    let time_type = field_type.tp().try_into()?;
+    DateTime::from_packed_u64(v, time_type, fsp, time_zone)
+}
+
+// TODO: Ideally we should not mark these functions as inline, since they are neither small, nor
+// hot enough. However, currently we have a super large `Result`, which must be inlined to avoid
```

Comment:
Do you mean the inline functions above? Looks like they are small.
BTW, to my experiment, inlining functions doesn't always avoid the cost of the clone. The compiler can optimize out this clone only if the function has Ok/Err branch only.

Reply:
Oh, this comment is outdated. I will remove it.